crates/ruff_python_formatter/src/expression/string.rs

```diff
@@ -0,0 +1,260 @@
+            c => QuoteStyle::try_from(c),
+        };
+
+        match style {
+            Ok(QuoteStyle::Single) => {
+                single_quotes += 1;
+            }
+            Ok(QuoteStyle::Double) => {
+                double_quotes += 1;
+            }
+            Err(_) => {}
+        }
+    }
+
+    if double_quotes > single_quotes {
+        QuoteStyle::Single
+    } else {
+        QuoteStyle::Double
+    }
+}
+
+struct StringQuotes {
+    triple: bool,
+    style: QuoteStyle,
+}
+
+impl StringQuotes {
+    fn parse(input: &str) -> Option<StringQuotes> {
+        let mut chars = input.chars();
+
+        let quote_char = chars.next()?;
+        let style = QuoteStyle::try_from(quote_char).ok()?;
+
+        let triple = chars.next() == Some(quote_char) && chars.next() == Some(quote_char);
+
+        Some(Self { triple, style })
+    }
+
+    const fn text_len(&self) -> TextSize {
+        if self.triple {
+            TextSize::new(3)
+        } else {
+            TextSize::new(1)
+        }
+    }
+}
+
+impl Format<PyFormatContext<'_>> for StringQuotes {
+    fn fmt(&self, f: &mut Formatter<PyFormatContext<'_>>) -> FormatResult<()> {
+        let quotes = match (self.style, self.triple) {
+            (QuoteStyle::Single, false) => "'",
+            (QuoteStyle::Single, true) => "'''",
+            (QuoteStyle::Double, false) => "\"",
+            (QuoteStyle::Double, true) => "\"\"\"",
+        };
+
+        text(quotes).fmt(f)
+    }
+}
+
+/// Adds the necessary quote escapes and removes unnecessary escape sequences when quoting `input`
+/// with the provided `style`.
+fn normalize_quotes(input: &str, style: QuoteStyle) -> (Cow<str>, ContainsNewlines) {
+    let mut output = String::new();
+
+    let preferred_quote = style.as_char();
+    let opposite_quote = style.opposite().as_char();
+    let mut last_index = 0;
+    let mut contains_newlines = ContainsNewlines::No;
+
+    let mut chars = input.char_indices().peekable();
+
+    while let Some((index, c)) = chars.next() {
+        if c == '\\' {
+            if let Some((_, next)) = chars.peek() {
+                if *next == opposite_quote {
+                    chars.next();
+                    // Remove the escape
+                    output.push_str(&input[last_index..index]);
+                    last_index = index + '\\'.len_utf8();
+                } else if *next == preferred_quote {
+                    // Keep the escape
+                    chars.next();
+                }
+            }
+        } else if c == preferred_quote {
+            // Escape the quote
+            output.push_str(&input[last_index..index]);
+            output.push('\\');
+            output.push(c);
+            last_index = index + preferred_quote.len_utf8();
+        } else if matches!(c, '\n' | '\r') {
+            contains_newlines = ContainsNewlines::Yes;
+        }
+    }
+
+    let normalized = if last_index == 0 {
```

Comment:
that's a neat trick, but i think it needs a comment somewhere at the top of the function